R/linkMultipleVariants.R

```diff
@@ -0,0 +1,294 @@
+                      "avePhredMinForward", "avePhredMinReverse",
+                      "constantMaxDistForward", "constantMaxDistReverse")
+    for (are in argsRecEqual) {
+        for (parm in paramsSeparate) {
+            if (!identical(parm[[are]], combinedDigestParams[[are]])) {
+                message("We recommend that the argument ", are,
+                        " is set to the same value in the ",
+                        "combined and all the separate runs")
+            }
+        }
+    }
+
+    ## Adding UMI counts after the processing may not be accurate - will
+    ## always use read counts
+    if (length(grep("U", paste0(combinedDigestParams$elementLengthsForward,
+                                combinedDigestParams$elementLengthsReverse))) > 0) {
+        warning("Aggregating UMI counts may not be accurate - will use ",
+                "read counts instead")
+    }
+
+    ## Get number and lengths of forward and reverse variable sequences
+    nbrFwdV <- sum(strsplit(combinedDigestParams$elementsForward, "")[[1]] == "V")
+    nbrRevV <- sum(strsplit(combinedDigestParams$elementsReverse, "")[[1]] == "V")
+    ## Check each of the separate runs
+    lengthsV <- unlist(lapply(paramsSeparate, function(parm) {
+        if (!parm$mergeForwardReverse) {
+            ## Don't merge - there can be only one V in total, and the length
+            ## must be given
+            els <- c(strsplit(parm$elementsForward, "")[[1]],
+                     strsplit(parm$elementsReverse, "")[[1]])
+            ell <- c(parm$elementLengthsForward, parm$elementLengthsReverse)
+            if (sum(els == "V") != 1) {
+                stop("Must have exactly one V in each separate run")
+            }
+            if (any(ell[which(els == "V")] == -1)) {
+                stop("The length of the variable sequence must be explicitly ",
+                     "specified in the separate runs")
+            }
+            ell[which(els == "V")]
+        } else {
+            ## Merge - there must be a V in both fwd and reverse, with the
+            ## same length
+            elsfwd <- strsplit(parm$elementsForward, "")[[1]]
+            elsrev <- strsplit(parm$elementsReverse, "")[[1]]
+            ellfwd <- parm$elementLengthsForward
+            ellrev <- parm$elementLengthsReverse
+            indexVfwd <- which(elsfwd == "V")
+            indexVrev <- which(elsrev == "V")
+            if (length(indexVfwd) != 1 || length(indexVrev) != 1) {
+                stop("Must have exactly one V in the forward and one V in ",
+                     "the reverse read in each separate run for ",
+                     "mergeForwardReverse=TRUE")
+            }
+            if ((ellfwd[indexVfwd] != ellrev[indexVrev]) ||
+                any(c(ellfwd[indexVfwd], ellrev[indexVrev]) == -1)) {
+                stop("The length of the variable sequence must be explicitly ",
+                     "specified and identical between the forward and reverse ",
```

Comment:
Probably a dumb question: Why do the forward and reverse Vs have to have identical length?

Reply:
My thought was that if they are to be merged, and we need to be able to infer how long the resulting V is to extract it from the concatenated string in the end, we might need to restrict ourselves to situations where we're essentially sequencing the same construct in the fwd and rev (if there's only one V in the fwd/rev, and they need to be merged with more relaxed conditions, `digestFastqs` will do the job). Maybe this is too restrictive though. And if not, we probably need to add a check that the fwd/rev variable sequence need to overlap completely.

Reply:
Right, that makes sense. I think it does not harm to be on the safe side here, at least until we have a use-case that needs less restrictive rules.

Reply:
Yes, I think we should do that.